File: 2182-construct-string-with-repeat-limit/2182-construct-string-with-repeat-limit.java
import java.util.*;

class Solution {
    public String repeatLimitedString(String s, int repeatLimit) {
        int[] freq = new int[26];
        for (char c : s.toCharArray()) freq[c - 'a']++;

        PriorityQueue<Character> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int i = 0; i < 26; i++) if (freq[i] > 0) pq.add((char) (i + 'a'));

        StringBuilder result = new StringBuilder();

        while (!pq.isEmpty()) {
            char largest = pq.poll();
            int idx = largest - 'a';
            int count = Math.min(repeatLimit, freq[idx]);
            result.append(String.valueOf(largest).repeat(count));
            freq[idx] -= count;

            if (freq[idx] > 0 && !pq.isEmpty()) {
                char nextLargest = pq.poll();
                result.append(nextLargest);
                freq[nextLargest - 'a']--;
                if (freq[nextLargest - 'a'] > 0) pq.add(nextLargest);
                pq.add(largest);
            }
        }

        return result.toString();
    }
}

File: 2182-construct-string-with-repeat-limit/2182-construct-string-with-repeat-limit.cpp
class Solution {
public:
    string repeatLimitedString(string s, int repeatLimit) {
        vector<int>frq(26);
        for(char c : s)frq[c-'a']++;
        vector<int>v;
        for(int i = 0 ; i < 26 ;i++)
            if(frq[i])v.push_back('a'+i);

        s= "";

        while(v.size()){ // a b c ,,,,, z 
            int lst = v.back() - 'a'; // 26
            v.pop_back();
            int mn = min(repeatLimit,frq[lst]); // 2
            frq[lst]-=mn;
            s+=string(mn,lst+'a'); ///zzccc

            if(v.size() && frq[lst]){
                int cur = v.back() - 'a';
                s.push_back('a'+cur);
                frq[cur]--;
                if(frq[cur]==0)v.pop_back();
                v.push_back(lst+'a');
            }

        }
        return s;

    }
};